File: backend/app/schemas/cohort.py
from __future__ import annotations

from pydantic import BaseModel
from typing import Any


class CRSConfigResponse(BaseModel):
    id: str
    components: list[dict[str, Any]]
    tier_thresholds: list[dict[str, Any]]
    tiebreaker_rules: list[dict[str, Any]]


class CRSConfigUpdate(BaseModel):
    components: list[dict[str, Any]] | None = None
    tier_thresholds: list[dict[str, Any]] | None = None
    tiebreaker_rules: list[dict[str, Any]] | None = None


class RecalculateRequest(BaseModel):
    patient_ids: list[str] | None = None


class RecalculateResponse(BaseModel):
    processed: int
    tier_changes: int


class CRSBreakdownComponent(BaseModel):
    raw: int | float
    weighted: float


class AssignmentRecord(BaseModel):
    id: str
    patient_id: str
    patient_name: str
    tier_number: int
    previous_tier: int | None
    crs_score: int
    crs_breakdown: dict[str, CRSBreakdownComponent]
    assignment_type: str
    reason: str | None
    assigned_at: str
    review_due_at: str | None


class AssignmentListResponse(BaseModel):
    items: list[AssignmentRecord]
    total: int
    page: int
    page_size: int
    pages: int


class TierDistribution(BaseModel):
    tier: int
    count: int


class TierDistributionResponse(BaseModel):
    distribution: list[TierDistribution]
    total: int

File: src/services/types/cohort.ts
export interface DashboardStats {
  total_patients: number;
  assigned: number;
  unassigned: number;
  pending_rescore: number;
  active_programs: number;
}

export interface CohortDistribution {
  cohort_id: string;
  cohort_name: string;
  cohort_color: string;
  count: number;
}

export interface AssignmentRecord {
  id: string;
  patient_id: string;
  patient_name: string;
  program_id: string;
  program_name: string | null;
  cohort_id: string;
  cohort_name: string;
  cohort_color: string;
  score: number | null;
  score_breakdown: Record<string, { raw: number; weighted: number }> | null;
  assignment_type: string;
  reason: string | null;
  previous_cohort_id: string | null;
  previous_cohort_name: string | null;
  pdc_worst: number | null;
  assigned_at: string;
  review_due_at: string | null;
}

export interface AssignmentListResponse {
  items: AssignmentRecord[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

export interface RecalculateResponse {
  events_created: number;
  scope: string;
}

export interface ScoringEventData {
  patient_name: string;
  score: number | null;
  cohort_id: string;
  cohort_name: string;
  cohort_color: string;
  assignment_type: string;
  program_id: string;
  program_name: string;
  assigned_at: string;
  review_due_at: string | null;
}

export interface SSEEvent {
  type: "batch_started" | "item_processed" | "item_failed" | "batch_complete";
  entity_id?: string;
  data: Record<string, unknown>;
}
